keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,546 @@
+
+        // Action Summary status
+        // Default is nil
+        // +optional
+        ActionSucceed *bool
+}
+
+// ContainerExecutionHookStatus represents the current state of a hook for a specific container in a pod
+type ContainerExecutionHookStatus struct {
+        // This field is required
+        PodName string
+
+        // This field is required
+        ContainerName string
+
+        // The current state of an hook action on a container
+        // +optional
+        HookActionStatus *HookActionStatus
+ }
+
+type HookActionStatus struct {
+        // If not set, it is nil, indicating Action has not started
+        // If set, it means Action has started at the specified time
+	// +optional
+        ActionTimestamp *int64
+
+        // If not set, it is nil, indicating Action is not complete on the specified container in the pod
+        // +optional
+        ActionSucceed *bool
+
+        // The last error encountered when running the action
+	// +optional
+        HookError *HookError
+}
+
+type HookError struct {
+        // Type of the error
+        // This is required
+        ErrorType ErrorType
+
+        // Error message
+        // +optional
+        Message *string
+
+        // More detailed reason why error happens
+        // +optional
+        Reason *string
+}
+
+type ErrorType string
+
+// More error types could be added, e.g., ExecutionForbidden,
+// ExecutionUnauthorized, ExecutionAlreadyInProgress, etc.
+const (
+        // The execution hook times out
+        ExecutionTimeout ErrorType = "ExecutionTimeout"
+
+        // The execution hook fails with an error
+        ExecutionFailwithError ErrorType = "ExecutionFailwithError"
+)
+```
+
+In the ExecutionHookStatus object, there is a list of ContainerExecutionHookStatus for all selected containers in the pods, each ContainerExecutionHookStatus represents the state of the hook on a specific container.
+
+The ActionSucceed field is a summary status field to indicate the final status of the Action in the hook after the commands are run on all selected containers. If not set, this field is nil, meaning the hook controller still needs to wait for the Action commands to finish running on all containers. If set, either true or false, it means the Action commands have finished running (or timeout) on all containers.
+
+
+Here is the definition of HookAction:
+
+```
+// HookAction describes action commands to run on pods/containers based
+// on specified policies. HookAction will be created by the user and
+// can be re-used later. Snapshot Controller will create ExecutionHooks
+// based on HookActions specified in the snapshot spec. For example,
+// two HookActions, preSnapshotExecutionHook and postSnapshotExecutionHook,
+// are expected in the snapshot spec.
+// HookAction does not contain information on pods/containers because those are
+// runtime info.
+// HookAction is namespaced
+type HookAction struct {
```

Comment:
you already rename PodAction to HookAction? But we have mentioned PodAction above

Reply:
My bad.  Will search and replace all the PodAction.

Reply:
I called for this renaming. `HookAction` sounds more consistent with the rest of the types/structs.

Reply:
We have been going back and forth about this name :(
The current thought is that this is all about ExecutionHook and its lifecycle management. The hookAction seems more relevant to this concept.

Reply:
I'm proposing to rename ExecutionHook to PodAction or Action and that (long-term) this be embedded into Pod.

Reply:
I also thought about this name actually. One reason we kind of keep using ExecutionHook is it sounds including more than just a action on a pod. Just a random thought, what if we will have execution hook on application level in the future which allows users to easily start some actions from application-level instead of individual pod?  Similar to snapshot, we have volume level snapshot and then application level snapshot. PodAction sounds very tied to the pod.